source/common/upstream/cluster_manager_impl.cc

```diff
@@ -149,19 +149,31 @@ void ClusterManagerInitHelper::maybeFinishInitialize() {
             secondary_init_clusters_.empty());
   if (!secondary_init_clusters_.empty()) {
     if (!started_secondary_initialize_) {
-      const auto type_url = Config::getTypeUrl<envoy::config::endpoint::v3::ClusterLoadAssignment>(
-          envoy::config::core::v3::ApiVersion::V2);
       ENVOY_LOG(info, "cm init: initializing secondary clusters");
       // If the first CDS response doesn't have any primary cluster, ClusterLoadAssignment
       // should be already paused by CdsApiImpl::onConfigUpdate(). Need to check that to
       // avoid double pause ClusterLoadAssignment.
-      if (cm_.adsMux() == nullptr || cm_.adsMux()->paused(type_url)) {
-        initializeSecondaryClusters();
-      } else {
-        cm_.adsMux()->pause(type_url);
-        Cleanup eds_resume([this, type_url] { cm_.adsMux()->resume(type_url); });
-        initializeSecondaryClusters();
+      std::unique_ptr<Cleanup> maybe_eds_resume_v2;
+      std::unique_ptr<Cleanup> maybe_eds_resume_v3;
+      if (cm_.adsMux()) {
+        const auto type_url_v2 =
+            Config::getTypeUrl<envoy::config::endpoint::v3::ClusterLoadAssignment>(
+                envoy::config::core::v3::ApiVersion::V2);
+        const auto type_url_v3 =
+            Config::getTypeUrl<envoy::config::endpoint::v3::ClusterLoadAssignment>(
+                envoy::config::core::v3::ApiVersion::V3);
+        if (!cm_.adsMux()->paused(type_url_v2)) {
```

Comment:
Do we need to be reference counting in pause/resume to avoid this kind of logic?

Reply:
i don't think i have the knowledge on how to do this

Reply:
addressed